include/ck_tile/core/arch/arch.hpp

```diff
@@ -60,13 +61,253 @@ enum struct memory_operation_enum : std::uint16_t
+CK_TILE_HOST_DEVICE constexpr bool is_rdna_arch_id(amdgcn_target_arch_id arch_id)
+{
+    return is_gfx11_arch_id(arch_id) || is_gfx12_arch_id(arch_id);
+}
+
+/*! @brief Returns true if the given arch_id maps to wave32 (RDNA) */
+CK_TILE_HOST_DEVICE constexpr bool is_wave32_arch_id(amdgcn_target_arch_id arch_id)
+{
+    return is_rdna_arch_id(arch_id);
+}
+
+/*! @brief Returns true if the given arch_id maps to wave64 (CDNA) */
+CK_TILE_HOST_DEVICE constexpr bool is_wave64_arch_id(amdgcn_target_arch_id arch_id)
+{
+    return is_cdna_arch_id(arch_id);
+}
+
+/*! @brief SFINAE enabler for target architecture if it is in the list of supported architectures
+ * @tparam TargetId The target architecture ID to check
+ * @tparam SupportedArchs The list of supported architecture IDs
+ */
+template <amdgcn_target_arch_id TargetId, amdgcn_target_arch_id... SupportedArchs>
+using enable_if_target_arch_id_t = std::enable_if_t<is_any_value_of(TargetId, SupportedArchs...)>;
+
+/*! @brief SFINAE enabler for target architecture if it is CDNA arch
+ * @tparam TargetId The target architecture ID to check
+ */
+template <amdgcn_target_arch_id TargetId>
+using enable_if_cdna_target_id_t = std::enable_if_t<is_cdna_arch_id(TargetId)>;
+
+/*! @brief SFINAE enabler for target architecture if it is CDNA arch
+ * @tparam TargetId The target architecture ID to check
+ */
+template <amdgcn_target_arch_id TargetId>
+using enable_if_rdna_target_id_t = std::enable_if_t<is_rdna_arch_id(TargetId)>;
+
+/*! @brief SFINAE enabler for target architecture if it is gfx9
+ * @tparam TargetId The target architecture ID to check
+ */
+template <amdgcn_target_arch_id TargetId>
+using enable_if_gfx9_target_id_t = std::enable_if_t<is_gfx9_arch_id(TargetId)>;
+
+/*! @brief SFINAE enabler for target architecture if it is gfx11
+ * @tparam TargetId The target architecture ID to check
+ */
+template <amdgcn_target_arch_id TargetId>
+using enable_if_gfx11_target_id_t = std::enable_if_t<is_gfx11_arch_id(TargetId)>;
+
+/*! @brief SFINAE enabler for target architecture if it is gfx12
+ * @tparam TargetId The target architecture ID to check
+ */
+template <amdgcn_target_arch_id TargetId>
+using enable_if_gfx12_target_id_t = std::enable_if_t<is_gfx12_arch_id(TargetId)>;
+
+/*! @brief SFINAE enabler for target architecture if it is wave32
+ * @tparam TargetId The target architecture ID to check
+ */
+template <amdgcn_target_arch_id TargetId>
+using enable_if_wave32_target_id_t = std::enable_if_t<is_wave32_arch_id(TargetId)>;
+
+/*! @brief SFINAE enabler for target architecture if it is wave64
+ * @tparam TargetId The target architecture ID to check
+ */
+template <amdgcn_target_arch_id TargetId>
+using enable_if_wave64_target_id_t = std::enable_if_t<is_wave64_arch_id(TargetId)>;
+
+/*! @brief Returns the amdgcn_target_arch_id of the current compiler pass
+ */
+CK_TILE_HOST_DEVICE constexpr auto get_target_arch_id()
+{
+    if constexpr(CK_TILE_ARCH_GFX908)
```

Comment:
Since this is explicitly checking the preprocessor, it's probably cleaner and more expressive  to use preprocessor macros that to dress this up in c++1x:


#if CK_TILE_ARCH_GFX908
    return amdgcn_target_arch_id::GFX908;
#elif CK_TILE_ARCH_GFX90A
    return amdgcn_target_arch_id::GFX90A;
...

#else
    return amdgcn_target_arch_id::HOST;
#endif

Reply:
One of my goals was to prevent preprocessor macros leaking out of config.hpp into other parts of the code. It might make more sense this way if we move to capture those CK_TILE_ARCH values into a struct instead like you mentioned below.